File: C++/Sorting/MergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> merge(vector<int> arr1,  vector<int> arr2)
{
    vector<int> arr3{};
    int i{},j{};
    while(i<arr1.size() && j<arr2.size())
    {
        if(arr1[i]<arr2[j])
        {
            arr3.push_back(arr1[i]);
            i++;
        }
        else
        {
            arr3.push_back(arr2[j]);
            j++;
        }
    }

    while(i<arr1.size())
    {
        arr3.push_back(arr1[i]);
        i++;
    }
    while(j<arr2.size())
    {
        arr3.push_back(arr2[j]);
        j++;
    }

    return arr3;
}

void print(vector<int> arr)
{
    for(int i{};i<arr.size();i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}
vector<int> sort(vector<int> arr)
{
    if(arr.size()<=1)
    {
        return arr;
    }
    else
    {
        int mid = arr.size()/2;
        vector<int> left = sort(vector<int>(arr.begin(),arr.begin()+mid));
        vector<int> right = sort(vector<int>(arr.begin()+mid,arr.end()));
        return merge(left, right);
    }
}
int main()
{
    vector<int> a{2,5,1,0,9,10};    
    vector<int> c = sort(a);
    print(a);
    print(c);
}

File: Python/Sorting/MergeSort.py
def merge(arr1:list[int],arr2:list[int])->list[int]:
    arr=[]
    i=0
    j=0
    while i<len(arr1) and j<len(arr2):
        if arr1[i]<arr2[j]:
            arr.append(arr1[i])
            i+=1
        else:
            arr.append(arr2[j])
            j+=1
            
    while i<len(arr1):
        arr.append(arr1[i])
        i+=1
    while j<len(arr2):
        arr.append(arr2[j])
        j+=1
    
    
    return arr


def sort(arr):
    
    if len(arr)<=1:
        return arr
    else:
        mid = len(arr)//2
        left = sort(arr[:mid])
        right = sort(arr[mid:])
        sortedArray = merge(left, right)
        return sortedArray

arr = [2,5,9,0,5,6,8,85,12,8,10]
print(arr)
arr = sort(arr)
print(arr)
            
     
    